src/content-claims/content-claims-blockstore.js

```diff
@@ -0,0 +1,379 @@
+  }
+}
+
+/**
+ * @implements {AbstractBlockStore}
+ */
+export class ContentClaimsBlockstore extends AbstractBlockStore {
+  /** @type {AbstractClaimsClient['read']} */
+  #read
+  /** @type {URL|undefined} */
+  #url
+  /** @type {import('../kv-bucket/api.js').KVBucketWithRangeQueries}  */
+  #carpark
+
+  /**
+   * @param {object} options
+   * @param {AbstractClaimsClient['read']} options.read
+   * @param {import('../kv-bucket/api.js').KVBucketWithRangeQueries} options.carpark - keys like `${cid}/${cid}.car` and values are car bytes
+   * @param {URL} [options.url]
+   */
+  constructor ({ read, url, carpark }) {
+    super()
+    this.#read = read
+    this.#url = url
+    this.#carpark = carpark
+    void /** @type {Blockstore} */ (this) // eslint-disable-line no-void
+  }
+
+  /**
+   * @param {UnknownLink} link
+   */
+  async has (link) {
+    const result = await claimsHas(this.#read, link, this.#url)
+    return result
+  }
+
+  /**
+   * @param {UnknownLink} link
+   */
+  async get (link) {
+    const claims = await this.#read(link, { serviceURL: this.#url })
+    // eslint-disable-next-line no-unreachable-loop
+    for await (const block of claimsGetBlock(claims, link, this.#carpark)) {
+      return block
+    }
+    throw errCode(new Error('not found'), 'ERR_NOT_FOUND')
+  }
+}
+
+/**
+ * @param {AbstractClaimsClient['read']} read
+ * @param {UnknownLink} link - link to answer whether the content-claims at `url` has blocks for `link`
+ * @param {URL} [serviceURL] - serviceURL to claims service to read from
+ */
+async function claimsHas (
+  read,
+  link,
+  serviceURL
+) {
+  const claims = await read(link, { serviceURL })
+  // @todo consider checking the claims themselves
+  return Boolean(claims.length)
+}
+
+/**
+ * @param {Map<string, Promise<IndexEntry|undefined>>} map - keys are CID strings
+ * @returns {import('../api.js').IndexMap & Index}
+ */
+function createIndexEntryMap (map = new Map()) {
+  return {
+    get: async (key) => {
+      return map.get(key.toString())
+    },
+    has: async (key) => {
+      const hasKey = map.has(key.toString())
+      return hasKey
+    },
+    set: async (key, value) => {
+      map.set(key.toString(), Promise.resolve(value))
+    }
+  }
+}
+
+/**
+ * get index for a link from a content claims client
+ * @param {AsyncIterable<import('../api.js').LocationClaim>|Iterable<import('../api.js').LocationClaim>} claims
+ * @param {UnknownLink} link - link to answer whether the content-claims at `url` has blocks for `link`
+ * @returns {AsyncGenerator<Uint8Array>}
+ */
+async function * fetchBlocksForLocationClaims (claims, link) {
+  /**
+   * @type {ReadableStream<{
+   *   claim: import('@web3-storage/content-claims/client/api').LocationClaim
+   *   location: string
+   *   response: Response
+   *   block: import('carstream/api').Block
+   * }>}
+   */
+  const locatedBlocks = createReadableStream(claims)
+    // claim -> [claim, location]
+    .pipeThrough(new TransformStream({
```

Comment:
this pipe syntax keeps tripping me up, feels verbose, maybe this could be something like

```js
.pipeThrough(transforms(
  async extractLocation (claim, controller) {
    for (const location of claim.location) { controller.enqueue([claim, location]) }
  },
  async fetchLocation ([claim, location], controller) {
    const response = await fetch(location, { headers: { accept: 'application/vnd.ipld.car' } })
    if (response.ok) {
      return controller.enqueue([claim, location, response])
    }
    throw new Error(`unexpected not-ok response ${response}`)
  },
  async readToBlocks ([claim, location, response], controller) {
    await response.body?.pipeThrough(new CARReaderStream()).pipeTo(new WritableStream({
      write: (block) => {
        // @todo validate that block.bytes hashes to `link`
        controller.enqueue({ claim, location, response, block })
      }
    }))
))
```

?

Reply:
where does `transforms` come from?

Reply:
I'm excited to use [stream.compose](https://nodejs.org/api/stream.html#streampipelinesource-transforms-destination-options) in node but also trying here to use web streams and have as much as possible be easy to move to browsers. lmk if you know of a good library with those properties but I'd prefer not to make one rn.